alacritty_terminal/src/tty/unix.rs

```diff
@@ -121,17 +118,30 @@ impl Pty {
     }
 }
 
-#[cfg(target_os = "macos")]
-fn default_shell(pw: &Passwd<'_>) -> Program {
-    let shell_name = pw.shell.rsplit('/').next().unwrap();
-    let argv = vec![String::from("-c"), format!("exec -a -{} {}", shell_name, pw.shell)];
-
-    Program::WithArgs { program: "/bin/bash".to_owned(), args: argv }
+fn default_shell(pw: &Passwd<'_>) -> String {
+    // Look for a shell in SHELL, then in passwd.
+    env::var("SHELL").unwrap_or_else(|_| pw.shell.to_owned())
 }
 
 #[cfg(not(target_os = "macos"))]
-fn default_shell(pw: &Passwd<'_>) -> Program {
-    Program::Just(env::var("SHELL").unwrap_or_else(|_| pw.shell.to_owned()))
+fn default_shell_command(pw: &Passwd<'_>) -> Command {
+    Command::new(default_shell(pw))
+}
+
+#[cfg(target_os = "macos")]
+fn default_shell_command(pw: &Passwd<'_>) -> Command {
+    // On macOS, use the `login` command to launch the shell so programs like `w`, `who`, and
+    // `logname` work.
+    let shell = default_shell(pw);
+    let shell_name = shell.rsplit('/').next().unwrap();
+    let mut login_command = Command::new("/usr/bin/login");
+    // -f bypasses authentication for the already-logged-in user
+    // -l tells login that this is not a login session and not to change directories to $HOME
+    // -p preserves the environment
+    // exec the shell with argv[0] prepended by `-` so it becomes a login shell anyway
+    let exec = format!("exec -a -{} {}", shell_name, shell);
+    login_command.args(["-flp", pw.name, "/bin/sh", "-c", &exec]);
+    login_command
```

Comment:
This is somewhat of a mess, the comment just squished between two lines without any whitespace above it, documenting the two following lines in opposite order.

The exec comment should be above the exec line, the flag comments should be above the args line. If this code is so intricate that it requires documentation above every single line, then that's perfectly fine.

Also what does `-l` actually do? Can you maybe post the `login --help` output? Having `-l` be `no-login` seems odd.

And also please format the exec comment to be capitalized with a full stop at the end and maybe change the args comment to use `-f: Bypasses authentication for the logged-in user.` as a comment format.

Reply:
Yes, I agree it's a little odd. Here is `man login` for the flags I'm using:

```
    -f      When a user name is specified, this option indicates that proper authentication has already been done and that no password need be requested.  This option
             may only be used by the super-user or when an already logged in user is logging in as themselves.

             With the -f option, an alternate program (and any arguments) may be run instead of the user's default shell.  The program and arguments follows the user
             name.

     -l      Tells the program executed by login that this is not a login session (by convention, a login session is signalled to the program with a hyphen as the first
             character of argv[0]; this option disables that), and prevents it from chdir(2)ing to the user's home directory.  The default is to add the hyphen (this is
             a login session).

     -p      By default, login discards any previous environment.  The -p option disables this behavior.
 ```

Reply:
And sure, I can re-order the comments. The order it is now was due to adding the `&exec` to the same `args` call rather than after it in its own `arg` call, per the other feedback.

Reply:
Updated.